Sources/SwiftProtobuf/ProtobufJSONEncoding.swift

```diff
@@ -21,16 +21,32 @@ struct ProtobufJSONEncodingVisitor: ProtobufVisitor {
     private var encoder = ProtobufJSONEncoder()
     var result: String {return encoder.result}
 
-    init() {}
+    private var nameResolver: (Int) -> String?
+    private var anyTypeURL: String?
+
+    init(message: ProtobufJSONMessageBase, anyTypeURL: String? = nil) throws {
+        self.nameResolver =
+            ProtoNameResolvers.jsonFieldNameResolver(for: message)
+        self.anyTypeURL = anyTypeURL
 
-    init(message: ProtobufJSONMessageBase) throws {
         try withAbstractVisitor {(visitor: inout ProtobufVisitor) in
             try message.traverse(visitor: &visitor)
         }
     }
 
     mutating func withAbstractVisitor(clause: (inout ProtobufVisitor) throws -> ()) throws {
         encoder.startObject()
+
+        // TODO: This is a bit of a hack that exists as a workaround to make the
+        // hand-written Any serialization work with the new design. We need to
+        // generate those WKTs instead of maintaining the hand-written ones,
+        // handle the special cases differently, and then remove this.
```

Comment:
> A swift extension on a struct can override/replace a method defined in a protocol extension.

That's a different situation than the one I was referring to, which was that methods generated *directly. However, when I wrote that I was forgetting that `serializeJSON` was implemented in a protocol extension rather than being generated into the struct, so it should still be possible to generate the remaining WKTs and then introduce an extension that overrides `serializeJSON` to do the right thing, which is good.

I'm not crazy about the fact that it allows people to arbitrarily extend their own custom types with a different JSON representation (that's against spec), but we'd have to overhaul the design more to lock that down and I'm not sure it's worth it.

> This is how Swift Protobuf allows end developers to override/replace the default hash and equality implementations, for instance.

Which we shouldn't be allowing either, but that's a separate issue. Equality of messages is well-defined by protobuf (see #97), and users should not be allowed to provide a different implementation that could produce different results than what the spec defines.